packages/dev/core/src/AudioV2/abstractAudio/subNodes/abstractAudioSubGraph.ts

```diff
@@ -0,0 +1,146 @@
+import type { Nullable } from "../../../types";
+import type { AbstractAudioNode, AbstractNamedAudioNode } from "../abstractAudioNode";
+import type { _AbstractAudioSubNode } from "./abstractAudioSubNode";
+
+/**
+ * Adds common sub graph functionality to an audio node.
+ *
+ * Audio nodes such as static sounds, streaming sounds, and buses can use audio sub graphs to process audio internally
+ * before sending it to connected downstream audio nodes. This is useful for applying effects, spatial audio, and other
+ * audio processing tasks common to multiple audio node classes.
+ *
+ * A key feature of audio sub graphs is their audio sub nodes are created asynchronously on demand so the minimum set
+ * of sub nodes are used at all times to save memory and CPU resources. The tradeoff is a small delay when first
+ * setting a property backed by a sub node. This delay is avoided by using the appropriate options to initialize the
+ * sub node on creation, e.g. `spatialEnabled` and `stereoEnabled`, or by setting any creation option backed by the
+ * sub node, e.g. `spatialPosition` and `stereoPan`.
+ *
+ * @internal
+ */
+export abstract class _AbstractAudioSubGraph {
+    private _createSubNodePromises: { [key: string]: Promise<_AbstractAudioSubNode> } = {};
+    private _subNodes: { [key: string]: Set<AbstractNamedAudioNode> } = {};
+
+    /**
+     * Executes the given callback with the named sub node, creating the sub node if needed.
+     *
+     * Note that `callback` is executed synchronously if the sub node already exists, otherwise the sub node is created
+     * asynchronously before `callback` is executed.
+     *
+     * @param name The name of the sub node
+     * @param callback The function to call with the named sub node
+     */
+    public callOnSubNode<T extends _AbstractAudioSubNode>(name: string, callback: (node: T) => void): void {
+        const node = this.getSubNode(name);
+        if (node) {
+            callback(node as T);
+            return;
+        }
+
+        const promise = this._createSubNodePromises[name] ?? this._createAndAddSubNode(name);
```

Comment:
Is it possible for this promise to fail? No code is catching errors, so if something were to happen, it will just disappear.

Reply:
Discussed offline and fixed